zebra-network/src/peer_set/candidate_set.rs

```diff
@@ -298,20 +301,20 @@ where
 
     /// Add new `addrs` to the address book.
     fn send_addrs(&self, addrs: impl IntoIterator<Item = MetaAddr>) {
-        let addrs = addrs
+        let addrs: Vec<MetaAddrChange> = addrs
             .into_iter()
             .map(MetaAddr::new_gossiped_change)
-            .map(|maybe_addr| {
-                maybe_addr.expect("Received gossiped peers always have services set")
-            });
+            .map(|maybe_addr| maybe_addr.expect("Received gossiped peers always have services set"))
+            .collect();
 
         // # Correctness
         //
-        // Briefly hold the address book threaded mutex, to extend
-        // the address list.
+        // Spawn address book accesses on a blocking thread,
+        // to avoid deadlocks (see #1976).
         //
         // Extend handles duplicate addresses internally.
-        self.address_book.lock().unwrap().extend(addrs);
+        let address_book = self.address_book.clone();
+        tokio::task::spawn_blocking(move || address_book.lock().unwrap().extend(addrs));
```

Comment:
TODO: return the spawned task, collect in a futures unordered, and run them before returning

```suggestion
        tokio::task::spawn_blocking(move || address_book.lock().unwrap().extend(addrs))
```